### PR TITLE
Remove old branches

### LINKED ISSUE #48: Add support for CI config removal
When removing CI config like in [here](https://github.com/openshift-knative/hack/commit/9e59c5542ac12bd6cc5e1b78d3917d6914b9a827), the job should:
* reflect the removal on the `openshift/release` fork and also remove them there, for the branch
* reflect the related image mirror file, when a complete version is removed

### PR DESCRIPTION
**Pull request by skonto**
- Fixes #48 for all repos besides S-O.
- Remove branches < 1.9 and variants < 4.11
- Creates the following changes:

> On branch sync-serverless-ci
> Your branch is up to date with 'master'.
> 
> Changes to be committed:
>   (use "git restore --staged <file>..." to unstage)
> 	renamed:    ci-operator/config/openshift-knative/eventing-hyperfoil-benchmark/openshift-knative-eventing-hyperfoil-benchmark-main__412.yaml -> ci-operator/config/openshift-knative/eventing-hyperfoil-benchmark/openshift-knative-eventing-hyperfoil-benchmark-main__414.yaml
> 	deleted:    ci-operator/config/openshift-knative/eventing-istio/openshift-knative-eventing-istio-release-v1.8__410.yaml
> 	deleted:    ci-operator/config/openshift-knative/eventing-istio/openshift-knative-eventing-istio-release-v1.8__413.yaml
> 	modified:   ci-operator/config/openshift-knative/eventing-kafka-broker/openshift-knative-eventing-kafka-broker-release-next__411.yaml
> 	modified:   ci-operator/config/openshift-knative/eventing-kafka-broker/openshift-knative-eventing-kafka-broker-release-next__413.yaml
> 	modified:   ci-operator/config/openshift-knative/eventing-kafka-broker/openshift-knative-eventing-kafka-broker-release-v1.11__411.yaml
> 	modified:   ci-operator/config/openshift-knative/eventing-kafka-broker/openshift-knative-eventing-kafka-broker-release-v1.11__413.yaml
> 	deleted:    ci-operator/config/openshift-knative/eventing-kafka-broker/openshift-knative-eventing-kafka-broker-release-v1.6__412.yaml
> 	deleted:    ci-operator/config/openshift-knative/eventing-kafka-broker/openshift-knative-eventing-kafka-broker-release-v1.6__48.yaml
> 	deleted:    ci-operator/config/openshift-knative/eventing-kafka-broker/openshift-knative-eventing-kafka-broker-release-v1.7__412.yaml
> 	deleted:    ci-operator/config/openshift-knative/eventing-kafka-broker/openshift-knative-eventing-kafka-broker-release-v1.7__48.yaml
> 	deleted:    ci-operator/config/openshift-knative/eventing-kafka-broker/openshift-knative-eventing-kafka-broker-release-v1.8__410.yaml
> 	deleted:    ci-operator/config/openshift-knative/eventing-kafka-broker/openshift-knative-eventing-kafka-broker-release-v1.8__413.yaml
> 	deleted:    ci-operator/config/openshift-knative/eventing/openshift-knative-eventing-release-v1.6__412.yaml
> 	deleted:    ci-operator/config/openshift-knative/eventing/openshift-knative-eventing-release-v1.6__48.yaml
> 	deleted:    ci-operator/config/openshift-knative/eventing/openshift-knative-eventing-release-v1.7__412.yaml
> 	deleted:    ci-operator/config/openshift-knative/eventing/openshift-knative-eventing-release-v1.7__48.yaml
> 	deleted:    ci-operator/config/openshift-knative/eventing/openshift-knative-eventing-release-v1.8__410.yaml
> 	deleted:    ci-operator/config/openshift-knative/eventing/openshift-knative-eventing-release-v1.8__413.yaml
> 	deleted:    ci-operator/config/openshift-knative/serving/openshift-knative-serving-release-v1.8__410.yaml
> 	deleted:    ci-operator/config/openshift-knative/serving/openshift-knative-serving-release-v1.8__413.yaml
> 	modified:   ci-operator/jobs/openshift-knative/eventing-hyperfoil-benchmark/openshift-knative-eventing-hyperfoil-benchmark-main-periodics.yaml
> 	modified:   ci-operator/jobs/openshift-knative/eventing-hyperfoil-benchmark/openshift-knative-eventing-hyperfoil-benchmark-main-postsubmits.yaml
> 	modified:   ci-operator/jobs/openshift-knative/eventing-hyperfoil-benchmark/openshift-knative-eventing-hyperfoil-benchmark-main-presubmits.yaml
> 	deleted:    ci-operator/jobs/openshift-knative/eventing-istio/openshift-knative-eventing-istio-release-v1.8-periodics.yaml
> 	deleted:    ci-operator/jobs/openshift-knative/eventing-istio/openshift-knative-eventing-istio-release-v1.8-postsubmits.yaml
> 	deleted:    ci-operator/jobs/openshift-knative/eventing-istio/openshift-knative-eventing-istio-release-v1.8-presubmits.yaml
> 	deleted:    ci-operator/jobs/openshift-knative/eventing-kafka-broker/openshift-knative-eventing-kafka-broker-release-v1.6-periodics.yaml
> 	deleted:    ci-operator/jobs/openshift-knative/eventing-kafka-broker/openshift-knative-eventing-kafka-broker-release-v1.6-postsubmits.yaml
> 	deleted:    ci-operator/jobs/openshift-knative/eventing-kafka-broker/openshift-knative-eventing-kafka-broker-release-v1.6-presubmits.yaml
> 	deleted:    ci-operator/jobs/openshift-knative/eventing-kafka-broker/openshift-knative-eventing-kafka-broker-release-v1.7-periodics.yaml
> 	deleted:    ci-operator/jobs/openshift-knative/eventing-kafka-broker/openshift-knative-eventing-kafka-broker-release-v1.7-postsubmits.yaml
> 	deleted:    ci-operator/jobs/openshift-knative/eventing-kafka-broker/openshift-knative-eventing-kafka-broker-release-v1.7-presubmits.yaml
> 	deleted:    ci-operator/jobs/openshift-knative/eventing-kafka-broker/openshift-knative-eventing-kafka-broker-release-v1.8-periodics.yaml
> 	deleted:    ci-operator/jobs/openshift-knative/eventing-kafka-broker/openshift-knative-eventing-kafka-broker-release-v1.8-postsubmits.yaml
> 	deleted:    ci-operator/jobs/openshift-knative/eventing-kafka-broker/openshift-knative-eventing-kafka-broker-release-v1.8-presubmits.yaml
> 	deleted:    ci-operator/jobs/openshift-knative/eventing/openshift-knative-eventing-release-v1.6-periodics.yaml
> 	deleted:    ci-operator/jobs/openshift-knative/eventing/openshift-knative-eventing-release-v1.6-postsubmits.yaml
> 	deleted:    ci-operator/jobs/openshift-knative/eventing/openshift-knative-eventing-release-v1.6-presubmits.yaml
> 	deleted:    ci-operator/jobs/openshift-knative/eventing/openshift-knative-eventing-release-v1.7-periodics.yaml
> 	deleted:    ci-operator/jobs/openshift-knative/eventing/openshift-knative-eventing-release-v1.7-postsubmits.yaml
> 	deleted:    ci-operator/jobs/openshift-knative/eventing/openshift-knative-eventing-release-v1.7-presubmits.yaml
> 	deleted:    ci-operator/jobs/openshift-knative/eventing/openshift-knative-eventing-release-v1.8-periodics.yaml
> 	deleted:    ci-operator/jobs/openshift-knative/eventing/openshift-knative-eventing-release-v1.8-postsubmits.yaml
> 	deleted:    ci-operator/jobs/openshift-knative/eventing/openshift-knative-eventing-release-v1.8-presubmits.yaml
> 	deleted:    ci-operator/jobs/openshift-knative/serving/openshift-knative-serving-release-v1.8-periodics.yaml
> 	deleted:    ci-operator/jobs/openshift-knative/serving/openshift-knative-serving-release-v1.8-postsubmits.yaml
> 	deleted:    ci-operator/jobs/openshift-knative/serving/openshift-knative-serving-release-v1.8-presubmits.yaml
> 	modified:   core-services/image-mirroring/knative/mapping_knative_knative-nightly_eventing-kafka-broker_quay
> 	modified:   core-services/image-mirroring/knative/mapping_knative_knative-v1.11_eventing-kafka-broker_quay
> 	deleted:    core-services/image-mirroring/knative/mapping_knative_knative-v1.6_eventing-kafka-broker_quay
> 	deleted:    core-services/image-mirroring/knative/mapping_knative_knative-v1.6_eventing_quay
> 	deleted:    core-services/image-mirroring/knative/mapping_knative_knative-v1.7_eventing-kafka-broker_quay
> 	deleted:    core-services/image-mirroring/knative/mapping_knative_knative-v1.7_eventing_quay
> 	deleted:    core-services/image-mirroring/knative/mapping_knative_knative-v1.8_eventing-istio_quay
> 	deleted:    core-services/image-mirroring/knative/mapping_knative_knative-v1.8_eventing-kafka-broker_quay
> 	deleted:    core-services/image-mirroring/knative/mapping_knative_knative-v1.8_eventing_quay
> 	deleted:    core-services/image-mirroring/knative/mapping_knative_knative-v1.8_serving_quay

### CODE REVIEW

**Comment by skonto** [comment]
Should I move Eventing to 1.14 instead of 1.13?

**Review by ReToCode — APPROVED** [review]
/lgtm
/approve
/hold I think it's fine with 4.14, but up to @pierDipi to confirm

**Comment by skonto** [comment]
@ReToCode @pierDipi gentle ping.

File: config/eventing-hyperfoil-benchmark.yaml
Comment (ReToCode) [inline config/eventing-hyperfoil-benchmark.yaml:7]: version:

**Review by ReToCode — CHANGES_REQUESTED** [review]

**Comment by ReToCode** [comment]
Just to not miss it, are we doing the eventing bump to 4.14 in a separate PR?

**Review by pierDipi — COMMENTED** [review]
This will need to be implemented in the controller logic https://github.com/openshift-knative/hack/issues/48, so we would need to "remove the YAML files before generating"

**Comment by skonto** [comment]
@pierDipi in order to complete this for each repo:

- controller needs to read the repo config and marks what needs to be deleted based on the branches not in config
- removes old yaml files
- proceeds with the new files
If that is so I could try implement it.

File: config/eventing-hyperfoil-benchmark.yaml
Comment (skonto) [reply config/eventing-hyperfoil-benchmark.yaml:7]: yeah needs rebase

**Comment by pierDipi** [comment]
> @pierDipi just to verify, in order to complete this task for each repo:
> 
> 1. controller needs to read the repo config and mark what needs to be deleted based on the branches not in config
> 2. removes old yaml files
> * proceeds with the new files generation
> 
> If that is so I could try implement it, so we can have proper clean up.

yes!

**Comment by pierDipi** [comment]
We can also simply "remove all the YAML files in the repository config directory using a regex like `*.yaml`", so combining (1) and (2) without "marking what needs to be deleted" as it's equivalent when combined with (3)

**Comment by skonto** [comment]
I will make it a bit more complex keeping the exclude stuff instead of removing everything. I will have a PR shortly.

File: pkg/prowgen/prowgen_test.go
Comment (skonto) [inline pkg/prowgen/prowgen_test.go:20]: Ideally we could only use the branch name only and pass an array instead of a map, but not that important I would say.

File: pkg/prowgen/prowgen.go
Comment (skonto) [inline pkg/prowgen/prowgen.go:405]: S-O not covered yet but we could as soon as we decide what branches we want to keep. For now leaving it untouched although if we remove the related midstream branches the corresponding S-O ones should follow.

**Comment by skonto** [comment]
@ReToCode @mgencur @pierDipi pls review.

**Review by ReToCode — COMMENTED** [review]
Hm, can you elaborate why we need that special handling in `getOldBranchCandidatePrefixes`? We do not have semver in the release branches anywhere, right? (we only have a xx.xx pattern) I don't seem to get why we need this.

Also the tests you added are a bit confusing, looking at the `branches` and the `want` I cannot really make sense of it. Why do we have totally different versions in `want`?

**Comment by skonto** [comment]
> Hm, can you elaborate why we need that special handling in getOldBranchCandidatePrefixes?

I keep the option of skipping branches as the code was before by respecting: `deleteConfigsIfNeeded`. I don't delete evrything.

> Also the tests you added are a bit confusing, looking at the branches and the want I cannot really make sense of it. Why do we have totally different versions in want?

It is a test that makes sure that only branches with pattern release-vX.Y are parsed and if let's say you have the lower version release-v2.2 anything before v2.2 is removed unless there is some condition that want us to keep the branch.

**Comment by ReToCode** [comment]
Thanks for explaining, I think I get your thoughts now. But I still feel is a bit complex for what we do. Wouldn't something similar to to [Pierangelos proposal](https://github.com/openshift-knative/hack/pull/90#issuecomment-1812099378) and respecting `ToRegexp(r.IgnoreConfigs.Matches)`  work, or am I missing something? So basically like:

- Loop over all existing folders+files and remove everything that is not not in the `IgnoreConfigs`
- Regenerate the config for everything that is still defined in the `config` files

So we should end up with the same but do not need to parse and compare versions.

**Comment by skonto** [comment]
closing in favor of #93

**Pull request closed without merging**
